src/foam/core/session/PersistentSessionDAO.js

```diff
@@ -0,0 +1,165 @@
+    'foam.core.logger.Logger',
+    'foam.core.logger.Loggers',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.dao.java.JDAO',
+    'foam.dao.Journal',
+    'foam.dao.MDAO',
+    'foam.dao.NullDAO',
+    'foam.dao.ProxyDAO',
+    'foam.lang.ContextAgent',
+    'foam.lang.ContextAgentTimerTask',
+    'foam.lang.X',
+    'java.util.HashMap',
+    'java.util.List',
+    'java.util.Map',
+    'java.util.Timer'
+  ],
+
+  javaCode: `
+  public PersistentSessionDAO(X x, DAO delegate) {
+    setX(x);
+    setDelegate(delegate);
+  }
+  `,
+
+  properties: [
+    {
+      name: 'persistentDAO',
+      class: 'foam.dao.DAOProperty',
+      javaFactory: 'return new MDAO(Session.getOwnClassInfo());'
+    },
+    {
+      name: 'mDAO',
+      class: 'foam.dao.DAOProperty'
+    },
+    {
+      name: 'journal',
+      class: 'FObjectProperty',
+      of: 'foam.dao.Journal'
+    },
+    {
+      name: 'anonymousCache',
+      class: 'Map',
+      javaFactory: 'return new HashMap();'
+    },
+    {
+      name: 'nullDAO',
+      class: 'foam.dao.DAOProperty',
+      javaFactory: 'return new NullDAO(getX(), Session.getOwnClassInfo());'
+    },
+    {
+      name: 'timerInterval',
+      class: 'Long',
+      value: 1, // 1000,
+      units: 'ms'
+    }
+  ],
+
+  methods: [
+    {
+      documentation: 'Load anonymous users',
+      name: 'start',
+      javaCode: `
+      X x = getX();
+      List<ServiceProvider> providers = (List) ((ArraySink) ((DAO) x.get("serviceProviderDAO")).select(new ArraySink())).getArray();
+      for ( ServiceProvider sp : providers ) {
+        long uid = sp.getAnonymousUser();
+        if ( uid != 0 ) {
+          getAnonymousCache().put(uid, true);
+        }
+      }
+
+      foam.core.logger.Logger logger = foam.core.logger.Loggers.logger(x, this, "start");
+      DAO dao = (DAO) x.get("localSessionDAO");
+      while ( dao instanceof ProxyDAO ) {
+        logger.info("dao", dao.getClass().getSimpleName());
+        if ( dao instanceof JDAO ) {
+          break;
+        }
+        dao = ((ProxyDAO) dao).getDelegate();
+      }
+      if ( dao instanceof JDAO ) {
+        setMDAO(((JDAO)dao).getDelegate());
+        setJournal(((JDAO)dao).getJournal());
+      } else {
+        throw new RuntimeException("PersistentSessionDAO JDAO not found");
+      }
+      `
+    },
+    {
+      name: 'put_',
+      javaCode: `
+        Session nu = (Session) getMDAO().put_(x, obj);
+        if ( nu.getUserId() == 0 ||
+             getAnonymousCache().containsKey(nu.getUserId()) ) {
+          return nu;
+        }
+
+        nu = (Session) getDelegate().put_(x, obj);
+
+        boolean save = false;
+        Session per = (Session) getPersistentDAO().find(nu.getId());
+        if ( per == null ) {
+          per = new Session();
+          per.copyFrom(nu);
+          save = true;
+        } else {
+           Map diff = per.diff(nu);
+           if ( diff.containsKey(Session.USER_ID.getName()) ||
```

Comment:
This feels flaky long term, maybe we should use the transient/storageTransient property properties on the fields we want to persist.